src/nspawn/nspawn.c

```diff
@@ -2523,6 +2524,23 @@ static int on_orderly_shutdown(sd_event_source *s, const struct signalfd_siginfo
         return 0;
 }
 
+static int on_sigchld(sd_event_source *s, const struct signalfd_siginfo *si, void *userdata) {
+        pid_t pid;
+
+        // Because we set PR_SET_CHILD_REAPER, orphan processes may be
+        // reparented to us.
+        pid = PTR_TO_PID(userdata);
+        if (pid == si->ssi_pid || (arg_clone_ns_flags & CLONE_NEWPID)) {
+                // The child has exited. nspawn can terminate now as well.
+                sd_event_exit(sd_event_source_get_event(s), 0);
+        } else {
+                // Orphan. Remove the zombie the from process table.
+                waitpid(si->ssi_pid, NULL, WNOHANG);
+        }
+
+        return 0;
+}
+
```

Comment:
That's a good catch! Thanks for the code sample too. I reused that with the exception that I added WEXITED to waitid() params so that we don't trip over SIGSTOP/SIGCONT as @evverx has pointed out. Please take another look.